src/adapters/adbutler.js

```diff
@@ -0,0 +1,148 @@
+/**
+ * @overview AdButler Prebid.js adapter.
+ * @author dkharton
+ */
+
+'use strict';
+
+var utils = require('../utils.js');
+var adloader = require('../adloader.js');
+var bidmanager = require('../bidmanager.js');
+var bidfactory = require('../bidfactory.js');
+
+var AdButlerAdapter = function AdButlerAdapter() {
+  
+  function _callBids(params) {
+    
+    var bids = params.bids || [],
+        callbackData = {},
+        zoneCount = {},
+        pageID = Math.floor(Math.random() * 10e6);
+
+    //Build and send bid requests
+    for (var i = 0; i < bids.length; i++) {
+      var bid = bids[i],
+        zoneID = utils.getBidIdParamater('zoneID', bid.params),
+        callbackID;  
+        
+      if(!(zoneID in zoneCount)){
+        zoneCount[zoneID] = 0;
+      }
+      
+      //build callbackID to get placementCode later  
+      callbackID = zoneID + '_' +zoneCount[zoneID];
+
+      callbackData[callbackID] = {};
+      callbackData[callbackID].bidId = bid.bidId;
+      
+      var adRequest = buildRequest(bid,zoneCount[zoneID],pageID);
+      zoneCount[zoneID]++;
+
+      adloader.loadScript(adRequest);
+    }
+      
+    //Define callback function for bid responses
+    $$PREBID_GLOBAL$$.adbutlerCB = function(aBResponseObject){
+        
+      var bidResponse = {},
+        callbackID = aBResponseObject.zone_id+'_'+aBResponseObject.place,
+        width = parseInt(aBResponseObject.width),
+        height = parseInt(aBResponseObject.height),
+        isCorrectSize = false,
+        isCorrectCPM = true,
+        CPM,minCPM,maxCPM,
+        bidObj = utils.getBidRequest(callbackData[callbackID].bidId);
+        
+      if (bidObj) {
+        
+        if(aBResponseObject.status === 'SUCCESS'){
+          CPM = aBResponseObject.cpm;
+          minCPM = utils.getBidIdParamater('minCPM',bidObj.params);
+          maxCPM = utils.getBidIdParamater('maxCPM',bidObj.params);
+
+          //Ensure response CPM is within the given bounds
+          if(minCPM !== '' && CPM < parseFloat(minCPM)){
+            isCorrectCPM = false;
+          }
+          if(maxCPM !== '' && CPM > parseFloat(maxCPM)){
+            isCorrectCPM = false;
+          }
+
+          //Ensure that response ad matches one of the placement sizes.  
+          utils._each(bidObj.sizes,function(size){
+            if(width === size[0] && height === size[1]){
+              isCorrectSize = true;
+            }
+          });
+
+          if(isCorrectCPM && isCorrectSize){
+
+            bidResponse = bidfactory.createBid(1);
```

Comment:
`createBid` can take `bidObj` as a second parameter here to set the response ID to the request ID. See #509 for background info